adr/0002-keycloak-gitlab.md

```diff
@@ -0,0 +1,162 @@
+
+#### Automated Group Management via Keycloak
+**Pros:**
+- Easier Access Management: Linking GitLab with Keycloak groups means when someone's group membership changes in Keycloak, it updates in GitLab too. This keeps access rights tight without extra effort.
+- Automatic Updates: Group memberships are updated automatically, saving a lot of manual work and ensuring no one is accidentally left in or out of a group.
+- Better Security: With automatic updates, you can be sure that everyone has the access they need and nothing more, keeping your projects and data safer.
+
+**Cons:**
+- Requires administering groups from Keycloak instead of GitLab. This may be more complex and not all users may have Keycloak group admin access. (However, this may scale better to large numbers of groups.)
+- Requires Careful Configuration: Setting up Keycloak to work correctly with GitLab requires careful planning and configuration.
+- Learning Curve: more complex to administer groups from Keycloak instead of GitLab. Another admin tool to learn and some complexity from not using GitLab directly.
+
+#### Manual Group Management
+**Pros:**
+- Simple to set up, likely works well for small GitLab installations
+- You're in Charge: Hand-managing groups means you make all the calls about who's in what group, giving you tight control over project access.
+**Cons:**
+- No integration with Keycloak groups required, less complexity
+
+### Instance Admin Group
+
+#### Automated Assignment via Keycloak Roles
+Note: requires using SAML authentication. See [SAML - Administrator groups](https://docs.gitlab.com/ee/integration/saml.html#administrator-groups) GitLab documentation.
+**Pros:**
+- Streamlines the process of granting administrative privileges to designated users.
+- Reduces the risk of unauthorized access by tightly controlling the assignment of admin roles.
+
+**Cons:**
+- It requires careful configuration to map Keycloak roles accurately to GitLab admin privileges.
+- Potential for misconfiguration, leading to unintended access rights.
+
+#### Manual Assignment of Admin Privileges
+**Pros:**
+- Direct control over who is granted administrative access to the GitLab instance.
+- Simple to manage on a small scale without the need for additional integration.
+
+**Cons:**
+- Labor-intensive, especially in larger organizations or those with frequent changes in administrative responsibilities.
+- Risk of oversight or delay in updating access rights, potentially impacting operations and security.
+
+## Decision
+Keycloak will be integrated with GitLab to enhance authentication and user management capabilities. The following decisions have been made:
+
+**_TBD_**
+1. **Authentication:** 
+2. **Provisioning Users:** 
+3. **Deprovisioning Users:**  
+4. **GitLab Group Integration:** 
+5. **Instance Admin Group:** 
```

Comment:
I took a stab at the Decision, but comments are welcome. This is a shift now from OIDC to SAML, but everything else can stay manual in the short term.

```suggestion
## Decision
Keycloak will be integrated with GitLab to enhance authentication and user management capabilities. The following decisions have been made:

1. **Authentication:** Use SAML. Select SAML instead of OIDC to support SAML group integration and SCIM in the future.
2. **Provisioning Users:** Automatic provisioning on first SSO login, users will only be created after their first login.  Long term, explore a user provisioning solution for all of UDS such as adding SCIM support to Keycloak.
3. **Deprovisioning Users:**  Short term, use manual provisioning and document a procedure for GitLab instance administrators. Long term, explore a user provisioning solution for all of UDS such as adding SCIM support to Keycloak.
4. **GitLab Group Integration:** Manual group creation should be supported. Additionally, test SAML Group Sync which should work with Keycloak.
5. **Instance Admin Group:** Manual admin creation should be supported. Additionally, test GitLab SAML Administrator groups which should work with Keycloak.
```